alacritty_terminal/src/selection.rs

```diff
@@ -174,26 +264,30 @@ impl Selection {
         let (start, end) = Selection::grid_clamp(start, end, lines, cols)?;
 
         let span = match *self {
-            Selection::Simple { ref region } => {
-                let (start_side, end_side) = if needs_swap {
-                    (region.end.side, region.start.side)
-                } else {
-                    (region.start.side, region.end.side)
-                };
-
-                self.span_simple(term, start, end, start_side, end_side)
-            },
-            Selection::Block { ref region } => {
-                let (start_side, end_side) = if needs_swap {
-                    (region.end.side, region.start.side)
-                } else {
-                    (region.start.side, region.end.side)
-                };
-
-                self.span_block(start, end, start_side, end_side)
-            },
-            Selection::Semantic { .. } => Selection::span_semantic(term, start, end),
-            Selection::Lines { .. } => Selection::span_lines(term, start, end),
+            Selection::Normal { .. } => {
+                // Simple Selection
+                let (start, end): (Anchor, Anchor) = self.span_simple(cols, start, end)?;
+
+                // Semantic expansion
+                let (start, end) = Selection::span_semantic(term, start, end);
+
+                // Line expansion
+                let (start, end) = Selection::span_line(cols, start, end);
```

Comment:
It's a bit unidiomatic and bad for performance to attempt every type of expansion regardless of what previous expansions have resulted in. A side will never require simple expansion, semantic expansion **and** line expansion, since it can only be one.

Reply:
I'll refactor this code to check the expansion type and then call the required expansion function.